JSON-Grammar-spec.md

```diff
@@ -0,0 +1,794 @@
+        {"r": "Number"},
+        {"p": ["(", {"r": "Expression"}, ")"], "ast": ["$", "/children/1"]}
+      ]
+    },
+    "Number": "/\\d+/",
+    "AddOp": {"u": ["+", "-"]},
+    "MulOp": {"u": ["*", "/"]}
+  },
+  "ast": {
+    "Number": ["num", ["$", "/raw"]]
+  }
+}
+```
+
+## Best Practices
+
+### Grammar Design
+
+1. **Start Simple**: Begin with basic rules and gradually add complexity
+2. **Use Meaningful Names**: Choose descriptive rule names that reflect their purpose
+3. **Leverage Shortcuts**: Use shorthand syntax where appropriate for cleaner grammars
+4. **Whitespace Handling**: Create dedicated whitespace rules with `ast: null` for clean ASTs
+5. **Left Recursion**: Avoid left-recursive rules; use right-recursion with lists instead
+
+### Performance Optimization
+
+1. **Order Alternatives**: Place most common alternatives first in union nodes
+2. **Minimize Backtracking**: Design grammars to reduce ambiguity
+3. **Atomic Groups**: Use terminal nodes for performance-critical patterns
+4. **Sample Data**: Provide sample strings for testing and validation
+
+### AST Design
+
+1. **Semantic Focus**: Include only semantically meaningful information in ASTs
+2. **Consistent Structure**: Maintain consistent AST node shapes across similar constructs
+3. **Type Safety**: Use clear, descriptive type names for AST nodes
+4. **Flatten Lists**: Transform complex nested structures into simpler forms
+
+### Error Handling
+
+1. **Debug Mode**: Use debug compilation for grammar development
+2. **Incremental Testing**: Test grammar rules individually before combining
+3. **Trace Analysis**: Leverage debug traces to understand parse failures
+4. **Sample Validation**: Verify grammars against known good and bad inputs
+
+### Debugging and Validation
+
+#### Debug Mode
+Enable debug mode during development to capture parsing traces:
+
+```javascript
+import {CodegenContext, ParseContext} from 'jit-parser';
+
+// Enable debug during compilation
+const debugCtx = new CodegenContext(true, true, true); // positions, ast, debug
+const parser = CodegenGrammar.compile(grammar, debugCtx);
+
+// Create trace collection
+const rootTrace = {pos: 0, children: []};
+const parseCtx = new ParseContext('input text', false, [rootTrace]);
+
+// Parse with debug trace
+const cst = parser(parseCtx, 0);
+
+// Print debug trace
+import {printTraceNode} from 'jit-parser';
+console.log(printTraceNode(rootTrace, '', 'input text'));
+```
+
+#### Grammar Visualization
+Use the built-in grammar printer to visualize grammar structure:
+
+```javascript
+import {GrammarPrinter} from 'jit-parser';
+
+const grammarString = GrammarPrinter.print(grammar);
+console.log(grammarString);
+```
+
+#### CST Inspection
+Print concrete syntax trees for debugging:
+
+```javascript
+import {printCst} from 'jit-parser';
+
+const parser = CodegenGrammar.compile(grammar);
+const ctx = new ParseContext('input', false);
+const cst = parser(ctx, 0);
+console.log(printCst(cst, '', 'input'));
+```
+
```

Comment:
Do not show code examples. This is a general purpose specification: it is programming language an library agnostic.

Instead, mention that debug traces can be captured in such and such way. And that debug trace nodes are such and such, and they form a tree in such and such way.